src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/TypeImportCompletionCacheEntry.cs

```diff
@@ -0,0 +1,243 @@
+
+            var isSameLanguage = Language == language;
+            var isInternalsVisible = originCompilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+            using var _ = ArrayBuilder<CompletionItem>.GetInstance(out var builder);
+
+            // PERF: try set the capacity upfront to avoid allocation from Resize
+            if (!isAttributeContext)
+            {
+                if (isInternalsVisible)
+                {
+                    builder.EnsureCapacity(ItemInfos.Length);
+                }
+                else
+                {
+                    builder.EnsureCapacity(PublicItemCount);
+                }
+            }
+
+            foreach (var info in ItemInfos)
+            {
+                if (!info.IsPublic && !isInternalsVisible)
+                {
+                    continue;
+                }
+
+                // Option to show advanced members is false so we need to exclude them.
+                if (hideAdvancedMembers && info.IsEditorBrowsableStateAdvanced)
+                {
+                    continue;
+                }
+
+                var item = info.Item;
+
+                if (isAttributeContext)
+                {
+                    // Don't show non attribute item in attribute context
+                    if (!info.IsAttribute)
+                    {
+                        continue;
+                    }
+
+                    // We are in attribute context, will not show or complete with "Attribute" suffix.
+                    item = GetAppropriateAttributeItem(info.Item, isCaseSensitive);
+                }
+
+                // C# and VB the display text is different for generics, i.e. <T> and (Of T). For simpllicity, we only cache for one language.
+                // But when we trigger in a project with different language than when the cache entry was created for, we will need to
+                // change the generic suffix accordingly.
+                if (!isSameLanguage && info.IsGeneric)
+                {
+                    // We don't want to cache this item.
+                    item = ImportCompletionItem.CreateItemWithGenericDisplaySuffix(item, genericTypeSuffix);
+                }
+
+                builder.Add(item);
+            }
+
+            return builder.ToImmutable();
+
+            static CompletionItem GetAppropriateAttributeItem(CompletionItem attributeItem, bool isCaseSensitive)
+            {
+                if (attributeItem.DisplayText.TryGetWithoutAttributeSuffix(isCaseSensitive: isCaseSensitive, out var attributeNameWithoutSuffix))
+                {
+                    // We don't want to cache this item.
+                    return ImportCompletionItem.CreateAttributeItemWithoutSuffix(attributeItem, attributeNameWithoutSuffix, CompletionItemFlags.Expanded);
+                }
+
+                return attributeItem;
+            }
+        }
+
+        public class Builder : IDisposable
```

Comment:
do you pass your builders around?  you coudl consider these being structs too.  if you do pass it around, yuo will need to pass by ref given the mutable publicItemCount.